# Build a CNN which exactly looks like VGG16.

In [4]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Input, Dense, Activation, Flatten, Conv2D, MaxPooling2D, InputLayer

In [5]:
def vgg16_clone():

    model = Sequential()

    # Block 1
    model.add(Conv2D(64, (3, 3), activation='relu', padding='same', input_shape=(224, 224, 3)))
    model.add(Conv2D(64, (3, 3), activation='relu', padding='same'))
    model.add(MaxPooling2D((2, 2), strides=2))

    # Block 2
    model.add(Conv2D(128, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(128, (3, 3), activation='relu', padding='same'))
    model.add(MaxPooling2D((2, 2), strides=2))

    # Block 3
    model.add(Conv2D(256, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(256, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(256, (3, 3), activation='relu', padding='same'))
    model.add(MaxPooling2D((2, 2), strides=2))

    # Block 4
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(MaxPooling2D((2, 2), strides=2))

    # Block 5
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(Conv2D(512, (3, 3), activation='relu', padding='same'))
    model.add(MaxPooling2D((2, 2), strides=2))

    # Flatten the output and add fully connected layers
    model.add(Flatten())
    model.add(Dense(4096, activation='relu'))
    model.add(Dense(4096, activation='relu'))
    model.add(Dense(1000, activation='softmax'))  # 1000 classes for ImageNet

    model.summary()

    return model


In [6]:
vgg16_clone()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_13 (Conv2D)              │ (None, 224, 224, 64)   │         1,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_14 (Conv2D)              │ (None, 224, 224, 64)   │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 112, 112, 64)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_15 (Conv2D)              │ (None, 112, 112, 128)  │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_16 (Conv2D)              │ (None, 112, 112, 128)  │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_6 (MaxPooling2D)  │ (None, 56, 56, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_17 (Conv2D)              │ (None, 56, 56, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_18 (Conv2D)              │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_19 (Conv2D)              │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_7 (MaxPooling2D)  │ (None, 28, 28, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_20 (Conv2D)              │ (None, 28, 28, 512)    │     1,180,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_21 (Conv2D)              │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_22 (Conv2D)              │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_8 (MaxPooling2D)  │ (None, 14, 14, 512)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_23 (Conv2D)              │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_24 (Conv2D)              │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_25 (Conv2D)              │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_9 (MaxPooling2D)  │ (None, 7, 7, 512)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 25088)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 4096)           │   102,764,544 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1000)           │     4,097,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

<Sequential name=sequential_1, built=True>

## Original VGG16

In [7]:
from tensorflow.keras.applications import VGG16

# Load the VGG16 model with pre-trained weights (on ImageNet)
vgg16_model = VGG16(include_top=True, weights='imagenet', input_shape=(224, 224, 3))

# Display the model's architecture summary
vgg16_model.summary()

553467096/553467096 ━━━━━━━━━━━━━━━━━━━━ 156s 0us/step


Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer_2 (InputLayer)      │ (None, 224, 224, 3)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv1 (Conv2D)           │ (None, 224, 224, 64)   │         1,792 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_conv2 (Conv2D)           │ (None, 224, 224, 64)   │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block1_pool (MaxPooling2D)      │ (None, 112, 112, 64)   │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv1 (Conv2D)           │ (None, 112, 112, 128)  │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_conv2 (Conv2D)           │ (None, 112, 112, 128)  │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block2_pool (MaxPooling2D)      │ (None, 56, 56, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv1 (Conv2D)           │ (None, 56, 56, 256)    │       295,168 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv2 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_conv3 (Conv2D)           │ (None, 56, 56, 256)    │       590,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block3_pool (MaxPooling2D)      │ (None, 28, 28, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv1 (Conv2D)           │ (None, 28, 28, 512)    │     1,180,160 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv2 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_conv3 (Conv2D)           │ (None, 28, 28, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block4_pool (MaxPooling2D)      │ (None, 14, 14, 512)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv1 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv2 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_conv3 (Conv2D)           │ (None, 14, 14, 512)    │     2,359,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ block5_pool (MaxPooling2D)      │ (None, 7, 7, 512)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 25088)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc1 (Dense)                     │ (None, 4096)           │   102,764,544 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ fc2 (Dense)                     │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ predictions (Dense)             │ (None, 1000)           │     4,097,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

# VGG16: A Detailed Explanation

---
VGG16 is a Convolutional Neural Network (CNN) architecture proposed by Karen Simonyan and Andrew Zisserman in their 2014 paper, *Very Deep Convolutional Networks for Large-Scale Image Recognition*. It is widely recognized for its simplicity, effectiveness, and excellent performance in image classification tasks, including the ImageNet Large Scale Visual Recognition Challenge (ILSVRC).

---
## 1. Architecture Overview

- **VGG** stands for **Visual Geometry Group**, which developed this architecture.
- **16** refers to the total number of weight layers in the network (13 convolutional layers + 3 fully connected layers).
- The model consists of:
  - Convolutional layers (with 3x3 filters).
  - Max-pooling layers (for down-sampling).
  - Fully connected layers (for final classification).

---
## 2. Detailed Layer-by-Layer Explanation

### Input Layer
- **Input Shape**: `(224, 224, 3)` (a 224x224 RGB image).
- Images must be resized to this shape before being fed into the network.
- Pixel values are normalized to ensure faster convergence during training.

---
### Convolutional Layers
- Convolutional layers extract spatial features from the image using small filters (kernels).
- **Filter Size**: Always 3x3 (the smallest size that still captures spatial features effectively).
- **Stride**: 1 (ensures no skipping of pixels during convolution).
- **Padding**: "same" (preserves spatial dimensions after convolution).
- **Number of Filters**:
  - Increases progressively across layers: 64 → 128 → 256 → 512 → 512.
  - This progression helps the network learn hierarchical features (from edges to complex textures and objects).
- **Activation Function**: ReLU (Rectified Linear Unit) is applied after each convolution to introduce non-linearity.

---
### Max-Pooling Layers
- **Purpose**: Reduces the spatial dimensions of the feature maps while retaining important features.
- **Kernel Size**: 2x2.
- **Stride**: 2.
- **Effect**:
  - Halves the width and height of the feature map after each pooling operation.
  - Reduces computational cost and prevents overfitting by discarding less important information.

---
### Fully Connected Layers
- The output of the convolutional layers (flattened into a 1D vector) is fed into these dense layers.
- **First FC Layer**:
  - 4096 neurons.
  - ReLU activation.
- **Second FC Layer**:
  - 4096 neurons.
  - ReLU activation.
- **Output Layer**:
  - Number of neurons = 1000 (one for each class in ImageNet).
  - Softmax activation function to output probabilities for each class.

---
## 3. Layer Block Structure

VGG16 is divided into **5 blocks**, each with a combination of convolutional and pooling layers:

| Block | Layers in Block                                                                                          | Output Dimensions (H x W x Channels) |
|-------|---------------------------------------------------------------------------------------------------------|---------------------------------------|
| Block 1 | 2 Conv2D (64 filters) + 1 MaxPooling                                                                   | 224 x 224 → 112 x 112 x 64           |
| Block 2 | 2 Conv2D (128 filters) + 1 MaxPooling                                                                  | 112 x 112 → 56 x 56 x 128            |
| Block 3 | 3 Conv2D (256 filters) + 1 MaxPooling                                                                  | 56 x 56 → 28 x 28 x 256              |
| Block 4 | 3 Conv2D (512 filters) + 1 MaxPooling                                                                  | 28 x 28 → 14 x 14 x 512              |
| Block 5 | 3 Conv2D (512 filters) + 1 MaxPooling                                                                  | 14 x 14 → 7 x 7 x 512                |
| FC     | Flatten → Dense (4096) → Dense (4096) → Dense (1000, softmax)                                           | Output: 1000 classes                 |

---
## 4. Key Design Principles

1. **Small Filters**:
   - All convolutional layers use 3x3 filters with stride 1.
   - Using small filters in deeper layers achieves the same receptive field as larger filters but with fewer parameters and better efficiency.
2. **Depth**:
   - The depth of VGG16 (16 layers) enables it to capture more complex patterns than shallower networks.
3. **Uniform Architecture**:
   - The consistent use of 3x3 filters and 2x2 max-pooling creates a simple, modular design.

---
## 5. Number of Parameters

- Total parameters: ~138 million.
- Most parameters are in the fully connected layers.
  - Conv layers: ~14 million parameters.
  - FC layers: ~124 million parameters.

---
## 6. Strengths of VGG16

- **Performance**:
  - Achieves high accuracy on image classification tasks.
- **Modularity**:
  - Simple, consistent design makes it easy to understand and implement.
- **Transfer Learning**:
  - Pre-trained weights on ImageNet are widely used for fine-tuning on other tasks.

---
## 7. Limitations of VGG16

1. **Computational Cost**:
   - Large model size (~138 million parameters) requires significant memory and computational resources.
   - Not ideal for deployment on resource-constrained devices.
2. **Overfitting**:
   - The high parameter count increases the risk of overfitting on smaller datasets.
3. **Slow Training**:
   - Due to the depth and parameter count, training VGG16 is time-intensive.

---
## 8. Applications

- **Image Classification**: Recognizing objects in images.
- **Feature Extraction**: Using convolutional layers as a feature extractor for tasks like object detection.
- **Transfer Learning**: Fine-tuning pre-trained VGG16 weights on domain-specific datasets.

---
## 9. Comparisons with Other Models

| Model      | Year | Parameters    | Key Features                   |
|------------|------|---------------|---------------------------------|
| AlexNet    | 2012 | 62 million    | First deep CNN for ImageNet.   |
| VGG16      | 2014 | 138 million   | Smaller filters, deeper layers.|
| ResNet     | 2015 | 25.5 million  | Introduced residual connections.|
| Inception  | 2015 | ~23 million   | Mixed architectures with modules.|

---
